src/OrleansRuntime/MultiClusterNetwork/MultiClusterOracle.cs

```diff
@@ -0,0 +1,766 @@
+                    var answer = (MultiClusterData) await remoteOracle.PushAndPull(localData.Current);
+
+                    // apply what we have learnt
+                    var delta = localData.ApplyDataAndNotify(answer);
+                    
+                    status.LastException = null;
+                    logger.Verbose("-{0} FullGossipWithSilo successful, answer={1}", id, answer);
+
+                    PushMyStatusToNewDestinations(delta);
+                }
+                catch (Exception e)
+                {
+                    logger.Warn(ErrorCode.MultiClusterNetwork_GossipCommunicationFailure,
+                        string.Format("-{0} FullGossipWithSilo {1} in cluster {2} failed", id, destinationSilo, destinationCluster), e);
+                    status.LastException = e;
+                }
+                status.LastUse = DateTime.UtcNow;
+                status.Pending = false;
+            };
+
+            // queue the push - we are not awaiting it!
+            status.Task = pushasync(status.Task);
+        }
+
+
+        private void InjectConfiguration(ref MultiClusterData deltas)
+        {
+            if (this.injectedConfig == null)
+                return;
+
+            var data = new MultiClusterData(this.injectedConfig);
+            this.injectedConfig = null;
+
+            if (logger.IsVerbose)
+                logger.Verbose("-InjectConfiguration {0}", data.Configuration.ToString());
+
+            var delta = this.localData.ApplyDataAndNotify(data);
+
+            if (!delta.IsEmpty)
+                deltas = deltas.Merge(delta);
+        }
+
+        private void InjectLocalStatus(bool isGateway, ref MultiClusterData deltas)
+        {
+            var myStatus = new GatewayEntry()
+            {
+                ClusterId = clusterId,
+                SiloAddress = Silo,
+                Status = isGateway ? GatewayStatus.Active : GatewayStatus.Inactive,
+                HeartbeatTimestamp = DateTime.UtcNow,
+            };
+
+            GatewayEntry existingEntry;
+
+            // do not update if we are reporting inactive status and entry is not already there
+            if (!this.localData.Current.Gateways.TryGetValue(Silo, out existingEntry) && !isGateway)
+                return;
+
+            // send if status is changed, or we are active and haven't said so in a while
+            if (existingEntry == null
+                || existingEntry.Status != myStatus.Status
+                || (myStatus.Status == GatewayStatus.Active
+                      && myStatus.HeartbeatTimestamp - existingEntry.HeartbeatTimestamp > this.randomizedResendActiveStatusAfter))
+            {
+                logger.Verbose2("-InjectLocalStatus {0}", myStatus);
+
+                // update current data with status
+                var delta = this.localData.ApplyDataAndNotify(new MultiClusterData(myStatus));
```

Comment:
`InjectLocalStatus`, `InjectConfiguration`, `DemoteLocalGateways`, `PushAndPull` and others update `this.localData` before even confirming that the operation was successfuly pushed to the targets/channels. If there is an error it probably won't retry that in the next sync, since the timestamp would be current, right?